src/project/config.rs

```diff
@@ -0,0 +1,337 @@
+
+use std::path::Path;
+use project::ProjectError;
+
+/// Trait implemented by the `config!` macro.
+pub trait Config where Self: Sized {
+    /// Loads a configuration structure from a file.
+    /// Defaults if the file fails in any way.
+    fn load<P: AsRef<Path>>(P) -> Self;
+
+    /// Loads a configuration structure from a file.
+    fn load_no_fallback<P: AsRef<Path>>(P) -> Result<Self, ProjectError>;
+
+    /// Writes a configuration structure to a file.
+    fn write<P: AsRef<Path>>(&self, P) -> Result<(), ProjectError>;
+}
+
+/// The `config!` macro allows defining configuration structures that can load in
+/// `.yml`/`.yaml` files.
+///
+/// It implements the [`serde::Serialize`], [`serde::Deserialize`], and [`Config`] traits.
+/// As well as the standard libraries `std::default::Default` trait for the defaulting fields.
+///
+/// It follows Rust's syntax for defining structures with one addition, defaulting values.
+/// If the file fails to specific fields of a configuration, then it will print out
+/// an error describing the issue and then load in this defaulting value.
```

Comment:
That's the same for every configuration format with optional keys; you just have to document it (which isn't a bad idea anyways, because you can explain what they're doing...). But let's here someone else's opinion.